### PR TITLE
Mypy fixes to DagRun, TaskInstance, and db utils

### PR DESCRIPTION
**Pull request by uranusjr**
Some more typing fixups found when implementing another feature, submitted separately.

### CODE REVIEW

File: airflow/utils/db.py [hunk 1]
```diff
@@ -882,12 +883,12 @@ def _move_dangling_task_data_to_new_table(session, source_table: "Table", target
...
 
 
-def check_task_tables_without_matching_dagruns(session) -> Iterable[str]:
+def check_task_tables_without_matching_dagruns(session: Session) -> Iterable[str]:
     import sqlalchemy.schema
     from sqlalchemy import and_, outerjoin
 
     metadata = sqlalchemy.schema.MetaData(session.bind)
-    models_to_dagrun = [TaskInstance, TaskReschedule]
+    models_to_dagrun: List[Any] = [TaskInstance, TaskReschedule]
```
Comment (uranusjr) [inline airflow/utils/db.py:891]: Technically we can type this (what we want is the `__tablename__` attribute from SQLAlchemy), but SQLAlchemy is so dynamic, adding accurate type here does not really help much, so I don’t really want to bother.

File: airflow/models/dagrun.py [hunk 2]
```diff
@@ -398,7 +398,9 @@ def generate_run_id(run_type: DagRunType, execution_date: datetime) -> str:
 
     @provide_session
     def get_task_instances(
-        self, state: Optional[Iterable[TaskInstanceState]] = None, session=None
+        self,
+        state: Optional[Iterable[Optional[TaskInstanceState]]] = None,
+        session: Session = NEW_SESSION,
```
Comment (kaxil) [inline airflow/models/dagrun.py:403]: In a separate PR or this (your choice) - can you add a small section in https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#coding-style-and-best-practices to let devs know when they should use `NEW_SESSION` vs `Session` for type hints

**Review by kaxil — APPROVED** [review]

File: airflow/models/dagrun.py [hunk 3]
```diff
@@ -398,7 +398,9 @@ def generate_run_id(run_type: DagRunType, execution_date: datetime) -> str:
 
     @provide_session
     def get_task_instances(
-        self, state: Optional[Iterable[TaskInstanceState]] = None, session=None
+        self,
+        state: Optional[Iterable[Optional[TaskInstanceState]]] = None,
```
Comment (ashb) [inline airflow/models/dagrun.py:402]: Isn't `None` a valid TaskInstanceState? https://github.com/apache/airflow/blob/820e836c4a2e45239279d4d71e1db9434022fec5/airflow/utils/state.py#L35

File: airflow/utils/db.py (hunk 1 above)
Comment (ashb) [reply airflow/utils/db.py:891]: 👍🏻 Also, as this isn't called anywhere from outside of this module it doesn't impact the typing of the rest of the code base.

**Review by potiuk — APPROVED** [review]

File: airflow/models/dagrun.py (hunk 3 above)
Comment (uranusjr) [reply airflow/models/dagrun.py:402]: I suspect subclassing from `str` is the problem here (but we probably can’t fix that without breaking a ton of things).

File: airflow/models/dagrun.py (hunk 2 above)
Comment (uranusjr) [reply airflow/models/dagrun.py:403]: Submitted as #20181.

**Pull request merged**